rmf_task/src/rmf_task/Priority.hpp

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef SRC__RMF_TASK__PRIORITY_HPP
+#define SRC__RMF_TASK__PRIORITY_HPP
+
+#include <memory>
+
+#include <rmf_utils/impl_ptr.hpp>
+
+namespace rmf_task {
+
+/// Implement this for a new prioritization scheme.
+class Priority
+{
+public:
+
+  /// Get the value of this priority
+  virtual std::size_t value() const = 0;
```

Comment:
What I'm about to say might sound strange, but we don't need to make any virtual functions for this "interface" class.

The `Priority` class can just be a blank base class, and the `CostCalculator` implementations can just dynamic cast the `Priority` objects into whatever type the `CostCalculator` is compatible with.